### ISSUE TITLE
Use HTML's "allowed to use" or permission delegation for allowvr

### ISSUE THREAD

**Issue by foolip** [w3c/webvr#86]
https://html.spec.whatwg.org/multipage/embedded-content.html#allowed-to-use

The language in https://w3c.github.io/webvr/#allowvr-attribute doesn't say to traverse all frames to the top-level frame, so it's easy to circumvent by just creating an iframe with the attribute on it.

As with https://github.com/w3c/mediacapture-main/issues/268#issuecomment-231339319 I wonder if permission delegation might work here.

**Comment by toji** [comment]
Thanks! I intended to bring up the idea of using permission delegation ([spec link](https://noncombatant.github.io/permission-delegation-api/) for easy reference) with the other implementors today. Seems like a more forward looking path than the `allowvr` attribute.

**Comment by foolip** [comment]
Yes, I am secretly hoping that `allowfullscreen` will be the only attribute of this sort ever shipped, but of course if you think an `allowvr` still makes the most sense in the end, then that's fine.

**Comment by toji** [comment]
My only concern is that it's not clear to me how committed the other implementing browser are to the permission delegation spec. If nobody strenuously objects, I'm happy to go that route.

**Comment by raymeskhoury** [comment]
Just a heads up that we're looking at getting this added to the initial version of [Feature Policy](https://docs.google.com/document/d/1k0Ua-ZWlM_PsFCFdLMa8kaVTo32PeNZ4G7FFHqpFx4E/edit) (the permission delegation successor) right now.

**Comment by ddorwin** [comment]
Feature Policy seems like the general direction. See also https://groups.google.com/a/chromium.org/d/msg/blink-dev/vM3RM1OQNFU/FcuMLjVfAgAJ for implementation issues with the iframe attribute.

In the meantime, please remove the `allowvr` text or mark it as under reconsideration / do not implement.

**Comment by ddorwin** [comment]
https://github.com/WICG/feature-policy/issues/49 tracks this on the Feature Policy side.

**Comment by cvan** [comment]
Just commenting so we don't forget that the spec needs to be updated to reflect the [Feature Policy](https://docs.google.com/document/d/1k0Ua-ZWlM_PsFCFdLMa8kaVTo32PeNZ4G7FFHqpFx4E/edit) dependency. This in scope for 1.1, 1.2?

Also, do we want to call the feature `vr` or `webvr`? I say `vr`, per `navigator.vr` (#116/#152). 

```html
// vr
<iframe enable="vr" src="example.html"></iframe>
<iframe featurepolicy='{"vr": ["*"]}' src="example.html"></iframe>

// or `webvr` …
<iframe enable="webvr" src="example.html"></iframe>
<iframe featurepolicy='{"webvr": ["*"]}' src="example.html"></iframe>
```

On a related note, something we brought up at the last/first W3C Workshop: permission to access WebVR may want to be managed by the user/UA. If so, I'd recommend adding `"vr"` to the [enum](https://w3c.github.io/permissions/#permission-registry) in the [Permissions API](https://w3c.github.io/permissions/). So, you'd invoke it like this:

```js
navigator.permissions.query({name: 'vr'}).then(result => {
  if (result.status === 'granted') {
    // Enter VR.
  } else {
    // Unable to enter VR.
  }
  result.addEventListener('change', () => {
    if (result.state === 'granted') {
      // Can enter VR.
    } else {
      // Unable to enter VR.
    }
  });
```

**Comment by clelland** [comment]
In the latest drafts for Feature Policy, you'd enable it in a cross-origin frame with something like:

    <iframe allow="vr" src="https://other.com/example.html">
    <iframe allow="webvr" src="https://other.com/example.html">

(depending on the final feature name), or by including an HTTP header like

    FeaturePolicy: {"vr": ["*"]}

in the parent document.

Same-origin frames would be able to use VR if the parent document can.

(That shouldn't be necessary to spell out in this doc, though.)

This spec, I think, will need to declare `"vr"` or `"webvr"` to be the keyword for the *WebVR* feature, and that its default allowlist is ["self"] -- which will enable the feature at the top-level, but block by default in cross-origin frames.

This spec will also need to declare when to call into allowed-to-use, and what to do if that algorithm returns `false` -- whether `navigator.vr.getDisplays()`  returns an empty array, or `navigator.vr.getAvailability` should return `false` as well, or something else.

**Comment by toji** [comment]
Thanks for the update @clelland! Do you have any examples of other specs that declare the keyword and default allow list, or is this new enough syntax that it hasn't propagated out to other specs yet? (Mostly just looking for some reference material)

**Comment by clelland** [comment]
It's new enough that other specs are just starting to look at how to
integrate with it; for now, I'm maintaining a list of features and the
rough wording that I'd expect to see eventually merged as an appendix to
the feature policy spec.

That spec is just live now, on https://wicg.github.io/feature-policy/ --
https://wicg.github.io/feature-policy/#defined-features is what I've been
using for external specs. Let me know if that works for you, or if we need
to come up with something more precise.

**Comment by cvan** [comment]
Thanks! See update here too: https://github.com/WICG/feature-policy/issues/38#issuecomment-289520799
